internal/cluster/cluster.go

```diff
@@ -140,6 +145,97 @@ func (c *clusterAPI) BuildAndSign(
 	return completedSuccessfully, nil
 }
 
+func (c *clusterAPI) ExpandAndReplaceKernelMappings(
+	ctx context.Context,
+	mcm *hubv1beta1.ManagedClusterModule,
+	cluster clusterv1.ManagedCluster) error {
+
+	modSpec := mcm.Spec.ModuleSpec
+	mod := &kmmv1beta1.Module{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      mcm.Name,
+			Namespace: c.namespace,
+		},
+		Spec: modSpec,
+	}
+
+	logger := log.FromContext(ctx)
+
+	kernelVersions, err := c.kernelVersions(cluster)
+	if err != nil {
+		return err
+	}
+
+	mappings := make([]kmmv1beta1.KernelMapping, 0, len(kernelVersions))
+
+	for _, kernelVersion := range kernelVersions {
+		kernelVersion := strings.TrimSuffix(kernelVersion, "+")
+
+		kernelVersionLogger := logger.WithValues(
+			"kernel version", kernelVersion,
+		)
+
+		mld, err := c.kernelAPI.GetModuleLoaderDataForKernel(mod, kernelVersion)
+		if err != nil {
+			kernelVersionLogger.Info("no suitable container image found; skipping kernel version")
+			continue
+		}
+
+		kernelVersionLogger.V(1).Info("Found a valid mapping",
+			"image", mld.ContainerImage,
+		)
+
+		mapping := c.mappingFromModuleLoaderData(mld)
+		mappings = append(mappings, mapping)
+
+		digest, err := c.imageDigestForModuleLoaderData(ctx, mld)
+		if err != nil {
+			kernelVersionLogger.Info("skipping image tag replacement: %v", err)
+			continue
+		}
```

Comment:
Should we do this before appending? We would not have to modify the latest mapping with `mappings[len(mappings)-1]` which feels unnecessary to me.

Reply:
I have updated it accordingly, I hope it's more readable now :)